metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java

```diff
@@ -122,8 +132,13 @@ public PartitionRegistration build() {
                 throw new IllegalStateException("You must set last known elr.");
             }
 
+            if (directories == null) {
+                directories = DirectoryId.unassignedArray(replicas.length);
+            }
```

Comment:
This makes sense to me -- set them as unassigned when they aren't specified.  Note that this could be the case due to the feature not being enabled.

However, I suspect we should we do the same thing in `public PartitionRegistration(PartitionRecord record)`, otherwise sometimes we can have `PartitionRegistration` records with non-null but "unassigned" directory IDs and sometimes we can have null.  It feels to me that  it would be easier to reason about if we always have something there.

Reply:
Good point. I've moved this to the underlying constructor to deal with both cases.